core/lib/zksync_core/src/metrics.rs

```diff
@@ -0,0 +1,186 @@
+//! Application-wide metrics.
+
+use vise::{Buckets, Counter, EncodeLabelSet, EncodeLabelValue, Family, Gauge, Histogram, Metrics};
+
+use std::{fmt, time::Duration};
+
+use zksync_dal::transactions_dal::L2TxSubmissionResult;
+use zksync_types::{aggregated_operations::AggregatedActionType, proofs::AggregationRound};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "stage")]
+pub(crate) enum InitStage {
+    HttpApi,
+    WsApi,
+    ContractVerificationApi,
+    StateKeeper,
+    EthWatcher,
+    EthTxAggregator,
+    EthTxManager,
+    DataFetcher,
+    Tree,
+    WitnessGenerator(AggregationRound),
+}
+
+impl fmt::Display for InitStage {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::HttpApi => formatter.write_str("http_api"),
+            Self::WsApi => formatter.write_str("ws_api"),
+            Self::ContractVerificationApi => formatter.write_str("contract_verification_api"),
+            Self::StateKeeper => formatter.write_str("state_keeper"),
+            Self::EthWatcher => formatter.write_str("eth_watcher"),
+            Self::EthTxAggregator => formatter.write_str("eth_tx_aggregator"),
+            Self::EthTxManager => formatter.write_str("eth_tx_manager"),
+            Self::DataFetcher => formatter.write_str("data_fetchers"),
+            Self::Tree => formatter.write_str("tree"),
+            Self::WitnessGenerator(round) => write!(formatter, "witness_generator_{round:?}"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum BlockL1Stage {
+    Saved,
+    Mined,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "stage", rename_all = "snake_case")]
+#[allow(clippy::enum_variant_names)]
+pub(crate) enum L1StageLatencyLabel {
+    UncommittedBlock,
+    UnprovedBlock,
+    UnexecutedBlock,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "stage")]
+pub(crate) enum BlockStage {
+    Sealed,
+    Tree,
+    MetadataCalculated,
+    MerkleProofCalculated,
+    L1 {
+        l1_stage: BlockL1Stage,
+        tx_type: AggregatedActionType,
+    },
+}
+
+impl fmt::Display for BlockStage {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Sealed => formatter.write_str("sealed"),
+            Self::Tree => formatter.write_str("tree"),
+            Self::MetadataCalculated => formatter.write_str("metadata_calculated"),
+            Self::MerkleProofCalculated => formatter.write_str("merkle_proof_calculated"),
+            Self::L1 { l1_stage, tx_type } => {
+                let l1_stage = match l1_stage {
+                    BlockL1Stage::Saved => "save", // not "saved" for backward compatibility
```

Comment:
It's actually spelled in 2 different ways: `eth_sender` uses "save", and house keeper uses "saved".